### ISSUE TITLE
Using standard rand() to roll dice makes die rolls too predictable  

### ISSUE THREAD

**Issue by msducheminjr** [statelesscode/nerd_dice#1]
In the method ```NerdDice.total_dice``` the normal top level ```rand()``` method is used.

```ruby
  def self.total_dice(number_of_sides, number_of_dice = 1, opts = {})
    total = 0
    number_of_dice.times do
      total += rand(number_of_sides) + 1 # the offending line
    end
    total += opts[:bonus] if Integer === opts[:bonus]
    total
  end
```
It would be better to use ```Random.new``` and call ```rand()``` on that number or ```SecureRandom.rand()```. That way the random number will get its own seed instead of using the top level object space fixed seed.

**Comment by msducheminjr** [comment]
There are performance considerations as well here. These are the results of a benchmark comparing ```rand()```, ```SecureRandom.rand()``` and ```Random.new.rand()```.

```ruby
Benchmark.bmbm do |x|
  # normal rand()
  x.report(":rand") do # standard rand()
    n.times { rand(1000) }
  end

  # SecureRandom.rand()
  x.report("Sec.rand") do 
    n.times { SecureRandom.rand(1000) }
  end
  
  # Random.new.rand()
  x.report("Random.new") do
    n.times { Random.new.rand(1000) }
  end
end
```

Results:
```bash
Rehearsal ----------------------------------------------
:rand        1.179966   0.000000   1.179966 (  1.180382)
Sec.rand     7.308806   3.131088  10.439894 ( 10.444647)
Random.new  90.571000   8.076307  98.647307 ( 98.695057)
----------------------------------- total: 110.267167sec

                 user     system      total        real
:rand        1.167842   0.000000   1.167842 (  1.168327)
Sec.rand     7.223330   3.096226  10.319556 ( 10.324173)
Random.new  90.460728   8.162781  98.623509 ( 98.675495)
```

Each is about an order of magnitude slower in exchange for more entropy. In  most cases, people aren't going to be running a casino using NerdDice and normal ```rand()``` is okay because you don't know the seed. The best solution may be to allow for this in configuration:

```ruby
NerdDice.configure do |config|
  # default to SecureRandom which has a decent trade-off between entropy and speed
  config.randomization_technique = :secure_random # [:rand, :secure_random, :random_new, :randomized]
  # only needed if :randomized is chosen from randomization_technique 
  config.randomization_technique_randomizer = :secure_random # [:rand, :secure_random, :random_new, nil]
  # other future config
end
```

This way the trade-off of speed vs. entropy can be left up to the use case of the user's specific project.

**Comment by msducheminjr** [comment]
Another option is ```Random.new``` once instead of on each iteration (which gets you off the generic rand and is the fastest of the four options).

```ruby
require 'benchmark'
require 'securerandom'

n = 5_000_000

Benchmark.bmbm do |x|
  # normal rand()
  x.report(":rand") do # standard rand()
    n.times { rand(1000) }
  end

  # SecureRandom.rand()
  x.report("Sec.rand") do 
    n.times { SecureRandom.rand(1000) }
  end

  # Random.new.rand()
  x.report("Random.new once") do
    gen = Random.new
    n.times { gen.rand(1000) }
  end

  
  # Random.new.rand() every time
  x.report("Random.new every time") do
    n.times { Random.new.rand(1000) }
  end
end
```

Results:
```bash
Rehearsal ---------------------------------------------------------
:rand                   1.137721   0.000100   1.137821 (  1.138270)
Sec.rand                6.352808   2.779168   9.131976 (  9.137953)
Random.new once         0.631958   0.000006   0.631964 (  0.632191)
Random.new every time  90.384581   7.790681  98.175262 ( 98.264619)
---------------------------------------------- total: 109.077023sec

                            user     system      total        real
:rand                   1.115316   0.000000   1.115316 (  1.115742)
Sec.rand                6.611016   2.782373   9.393389 (  9.399656)
Random.new once         0.617824   0.000000   0.617824 (  0.618098)
Random.new every time  90.323462   7.461265  97.784727 ( 97.913637)
```

**Comment by msducheminjr** [comment]
Configuration from #8 allows for randomization technique configuration to be set. Now we just need to implement it.

**Comment by msducheminjr** [comment]
More speed research/alternatives/analysis:

```ruby
require 'benchmark'
require 'securerandom'

n = 5_000_000

Benchmark.bm do |x|
  # normal rand()
  x.report("normal_rand") do # standard rand()
    n.times { rand(1000) }
  end

  # Random.rand()
  x.report("Random.rand") do # standard rand()
    n.times { Random.rand(1000) }
  end

  # Kernel.rand()
  x.report("Kernel.rand") do # standard rand()
    n.times { Random.rand(1000) }
  end

  # SecureRandom.rand()
  x.report("Sec.rand") do 
    n.times { SecureRandom.rand(1000) }
  end

  # Random.new.rand()
  x.report("Random.new once") do
    gen = Random.new
    n.times { gen.rand(1000) }
  end

  # Random.new.rand() every time
  x.report("Random.new every time") do
    n.times { Random.new.rand(1000) }
  end
end
```
Results:
```
        user     system      total        real
normal_rand  1.143830   0.000001   1.143831 (  1.144303)
Random.rand  0.571916   0.000001   0.571917 (  0.572363)
Kernel.rand  0.575969   0.000002   0.575971 (  0.576181)
Sec.rand  6.638272   2.743271   9.381543 (  9.386943)
Random.new once  0.623213   0.000000   0.623213 (  0.623485)
Random.new every time 91.270428   7.424343  98.694771 ( 98.773597)
```

Normal rand is slower because it is traveling up the stack until it finds a method it can call. When you do that every iteration 5 million times, it adds up.  ```Random.rand()``` and ```Kernel.rand()``` both share the same seed and are about equal in terms of speed. Random.rand() is less forgiving about zero or negative arguments.

**Comment by msducheminjr** [comment]
Resolved by PR #18
